File: LeetCode/Medium/2410-Maximum_Matching_Of_Players_With_Trainers/2410-MaximumMatchingOfPlayersWithTrainers.js
/**
 * @param {number[]} players
 * @param {number[]} trainers
 * @return {number}
 */
var matchPlayersAndTrainers = function (players, trainers) {
    let n = players.length;
    let m = trainers.length;

    players.sort((a, b) => a - b);
    trainers.sort((a, b) => a - b);

    let j = 0;
    let count = 0;

    for (let i = 0; i < n; i++) {
        while (j < m && players[i] > trainers[j])
            j++;

        if (j < m) {
            count++;
            j++;
        } else
            break;
    }

    return count;
};

File: LeetCode/Medium/2410-Maximum_Matching_Of_Players_With_Trainers/2410-MaximumMatchingOfPlayersWithTrainers.ex
defmodule Solution do
  @spec match_players_and_trainers(players :: [integer], trainers :: [integer]) :: integer
  def match_players_and_trainers(players, trainers) do
    sorted_players = Enum.sort(players)
    sorted_trainers = Enum.sort(trainers)

    match_players(sorted_players, sorted_trainers, 0)
  end

  defp match_players([], _trainers, count), do: count

  defp match_players(_players, [], count), do: count

  defp match_players([curr_player | rest_p], [curr_trainer | rest_t], count) do
    if curr_player <= curr_trainer do
      match_players(rest_p, rest_t, count + 1)
    else
      match_players([curr_player | rest_p], rest_t, count)
    end
  end
end
